be/src/io/cache_input_stream.cpp

```diff
@@ -0,0 +1,99 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+#include "io/cache_input_stream.h"
+
+#include <fmt/format.h>
+
+#include "block_cache/block_cache.h"
+#include "util/hash_util.hpp"
+#include "util/runtime_profile.h"
+
+namespace starrocks::io {
+
+CacheInputStream::CacheInputStream(const std::string& filename, std::shared_ptr<SeekableInputStream> stream)
+        : _filename(filename), _stream(stream), _offset(0) {
+    // _cache_key.resize(8);
+    // char* data = _cache_key.data();
+    // uint64_t hash_value = HashUtil::hash64(filename.data(), filename.size(), 0);
+    // memcpy(data, &hash_value, sizeof(hash_value));
+    _cache_key = _filename;
+    _buffer.reserve(BLOCK_SIZE);
+    _size = _stream->get_size().value();
+}
+
+#ifdef WITH_BLOCK_CACHE
+StatusOr<int64_t> CacheInputStream::read(void* out, int64_t count) {
+    BlockCache* cache = BlockCache::instance();
+    int64_t end = _offset + count;
+    int64_t start_block_id = _offset / BLOCK_SIZE;
+    int64_t end_block_id = (end - 1) / BLOCK_SIZE;
+
+    char* p = static_cast<char*>(out);
+    for (int64_t i = start_block_id; i <= end_block_id; i++) {
+        int64_t off = i * BLOCK_SIZE;
+        int64_t size = std::min(BLOCK_SIZE, end - off);
+        int64_t load_size = std::min(BLOCK_SIZE, _size - off);
+
+        VLOG_FILE << "[CacheInputStream] offset = " << _offset << ", end = " << end << ", block_id = " << i
+                  << ", off = " << off << ", size = " << size << " , load_size = " << load_size;
+
+        StatusOr<size_t> res;
+        char* src = nullptr;
+        {
+            SCOPED_RAW_TIMER(&_stats.read_cache_ns);
+            _stats.read_cache_count += 1;
+            res = cache->read_cache(_cache_key, off, load_size, _buffer.data());
+            src = _buffer.data();
+            // TODO: Replace the above with a safe zero copy interface
+            //st = cache->read_cache_zero_copy(_cache_key, off, load_size, (const char**)&src);
+        }
+        if (res.status().is_not_found()) {
+            RETURN_IF_ERROR(_stream->read_at_fully(off, _buffer.data(), load_size));
```

Comment:
say `_block_size = 256B`, stream offset = 0. Caller calls two reads
```
read(out, 100);
read(out, 300);
```
seems the cache data will be corrupted in such case.

Reply:
> say `_block_size = 256B`, stream offset = 0. Caller calls two reads
> 
> ```
> read(out, 100);
> read(out, 300);
> ```
> 
> seems the cache data will be corrupted in such case.

Also violates the read align restriction and will return some invalid argument error message.

Reply:
it is ok to force alignment in cache interface, but it doesn't make sense to force stream read/write interface also align with block size, it makes the stream hard to use for caller.

Reply:
> it is ok to force alignment in cache interface, but it doesn't make sense to force stream read/write interface also align with block size, it makes the stream hard to use for caller.

Maybe I misunderstood your thoughts before, I mean the cache interface is align restricted, but the stream interface are not subject to this constrait. For your cases, the first read will read the first block data from remote, cache it and copy the head 100 bytes to the out buffer. Then the second read will be aligned to block size by stream, and read head 156 bytes from first block cache and then read the second block from remote, cache it and read other (300 - 156) bytes from the second block cache.

Reply:
ok. my mistake to read line 34. you are using the file size _size, not the requested size to read parent stream, it is ok then.